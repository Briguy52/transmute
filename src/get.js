import curry from "./curry";
import { Iterable } from "immutable";

/**
 * Retrieve the value at `key` from `subject`.
 *
 * @example
 * // returns 1
 * get('one', Map({one: 1, two: 2, three: 3}))
 *
 * @param  {any} key to lookup in `subject`.
 * @param  {Iterable|Object} subject in which to look up `key`.
 * @return {any} the value at `key`.
 */
function get(key, subject) {
  if (subject === null || subject === undefined) {
    return undefined;
  }
  if (Iterable.isIterable(subject)) {
    return subject.get(key);
  }
  return subject[key];
}

export default curry(get);
