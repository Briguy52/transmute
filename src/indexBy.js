import _keyedEquivalent from './internal/_keyedEquivalent';
import _reduce from './internal/_reduce';
import _set from './internal/_set';
import curry from './curry';
import { Iterable, Map, OrderedMap } from 'immutable';

function indexBy(keyMapper, subject) {
  return _reduce(
    _keyedEquivalent(subject),
    (acc, v, k) => _set(keyMapper(v, k, subject), v, acc),
    subject
  );
}

/**
 * Create a Map, or OrderedMap from `subject` with a key for each item
 * returned by `keyMapper`.
 *
 * @example
 * indexBy(get('id'), List.of({id: 123}, {id: 456}))
 * // returns Map { 123: {id: 123}, 456: {id: 456} }
 *
 * @param  {Function} keyMapper generates keys for each item
 * @param  {Iterable} subject to index
 * @return {KeyedIterable}
 */
export default curry(indexBy);
