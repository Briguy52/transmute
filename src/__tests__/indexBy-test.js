import { List, Map } from "immutable";
import indexBy from "../indexBy";
import get from "../get";

describe("transmute/indexBy", () => {
  const getId = get("id");
  const indexById = indexBy(getId);

  it("indexes an Array", () => {
    const input = [Map({ id: "123" }), Map({ id: "456" })];
    expect(indexById(input)).toMatchSnapshot();
  });

  it("indexes a List", () => {
    expect(
      indexById(List.of(Map({ id: "123" }), Map({ id: "456" })))
    ).toMatchSnapshot();
  });

  it("indexes a Map", () => {
    expect(
      indexById(
        Map({
          one: Map({ id: "123" }),
          two: Map({ id: "456" })
        })
      )
    ).toMatchSnapshot();
  });

  it("indexes an Object", () => {
    expect(
      indexById({
        one: { id: "123" },
        two: { id: "456" }
      })
    ).toMatchSnapshot();
  });
});
