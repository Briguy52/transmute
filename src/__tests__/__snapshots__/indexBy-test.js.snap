// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/indexBy indexes a List 1`] = `
Immutable.OrderedMap {
  "123": Immutable.Map {
    "id": "123",
  },
  "456": Immutable.Map {
    "id": "456",
  },
}
`;

exports[`transmute/indexBy indexes a Map 1`] = `
Immutable.Map {
  "123": Immutable.Map {
    "id": "123",
  },
  "456": Immutable.Map {
    "id": "456",
  },
}
`;

exports[`transmute/indexBy indexes an Array 1`] = `
Immutable.OrderedMap {
  "123": Immutable.Map {
    "id": "123",
  },
  "456": Immutable.Map {
    "id": "456",
  },
}
`;

exports[`transmute/indexBy indexes an Object 1`] = `
Immutable.OrderedMap {
  "123": Object {
    "id": "123",
  },
  "456": Object {
    "id": "456",
  },
}
`;
